### ISSUE TITLE
Feature Request

### ISSUE THREAD

**Issue by clive-g-brown** [arrayfire/arrayfire#2878]
Metal Backend.

**Comment by vade** [comment]
Seconded!

**Comment by zhiyuanzhai** [comment]
+1.

**Comment by jpp46** [comment]
+1
I would love to be able to use array fire on m1 mac gpu (maybe we could get a neural engine backend too somehow)

**Comment by 9prady9** [comment]
We had a duplicate request #3123  Responding to that here to avoid a scattered conversation.

Based on what I, personally, have checked out from Metal [specification][1] and [documentation][2], it seems like Metal has most of the required ingredients to be a valid backend of ArrayFire. Having said that, Metal is tied down to Apple platform where as all of our current backends are cross platform. Apart from, I am not sure if we can achieve feature parity with our current backends which is very important for us. To figure out if such feature parity is possible, more detailed analysis of metal framework needs to be carried to understand how much of our API can be supported right away.

Regarding https://github.com/arrayfire/arrayfire/issues/3123#issuecomment-830682884 @BA8F0D39 , I personally don't know what's the future of Metal on Apple platforms. But what you said about cross-platform nature, that is true. Regarding Vulkan being a compute successor to OpenCL, I don't think that is true. In fact, OpenCL far more flexible and useful API than Vulkan from scientific computing perspective. Today's Vulkan is at best, OpenGL compute shaders on steroids - that's what I gathered from it's [specification][3].

@atolkachiov As far as, the decision to provide a Apple only backend for ArrayFire, we haven't looked into it yet. Above all, from the language stand point, as @BA8F0D39  has pointed out, ArrayFire is written in C++ with C, C++, Rust, Python and other APIs. Having to write an entire code base in totally different language is a gargantuan task.

Putting together all of the above information, I feel like ArrayFire API for Metal is separate product all together.

[1]: https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf
[2]: https://developer.apple.com/documentation/metalperformanceshaders/
[3]: https://www.khronos.org/registry/vulkan/specs/1.2/html/vkspec.html#pipelines-compute

**Comment by BA8F0D39** [comment]
@9prady9 

![alt text](https://www.khronos.org/assets/uploads/apis/2020-opencl-api-overview-5_3.jpg "OpenCL and Vulkan")


https://www.khronos.org/opencl/
Khronos Group says that OpenCL kernels can run be compiled to the intermediate language SPIR-V.
SPIR-V can run on the Vulkan Compute runtime.
Vulkan Compute runs on Apple's metal.
https://github.com/KhronosGroup/MoltenVK


Theoretically, arrayfire can run Apple's metal without any major modifications.

**Comment by 9prady9** [comment]
@BA8F0D39 Yes, I forgot to mention these - thanks for bringing those up as well.

I am definitely not an expert on Vulkain, but from I read in Vulkan specification it seemed like Vulkan compute is a static compute shader style feature much like OpenGL compute shader but with some additional newer things like share memory etc. Not everything that ArrayFire needs especially JIT is possible using Vulkan API. https://github.com/google/clspv looks promising but even that provides only a subset of OpenCL C, so it would really would depend on how much of what ArrayFire needs is available in this subset of OpenCL C provided by clspv.

Overall theoretically either with couple of layers of abstractions or something like clspv, ArrayFire could run without modifications on Apple, eventually but this is a very long shot I think - for any production deployable code.

**Comment by atolkachiov** [comment]
I agree the future of Metal is unclear, and without native C++ API headers is might be not worth putting effort into its support now.  

Moreover, there is Neural Engine on all new apple chips which is physically different device than GPU which is rumoured to provide much higher performance for ML tasks. But there is no publicly available low-level API besides very high level [CoreML](https://developer.apple.com/documentation/coreml) wich decides where to execute models - either on Neural Engine TPU or Metal Performance Shaders for GPU or BNNS which runs on CPU.

**Comment by galad87** [comment]
Metal is the only available 3d api on Apple devices, so it's future it's not that unclear, and it's updated every year, what @BA8F0D39  said in the other issue is false.

The API is in Objective-C, but Objective-C can be called from C, C++, and Swift, and there are some open source C++ wrappers out there.

This said, I don't know if Metal has all the features you need for ArrayFire.

**Comment by vade** [comment]
Small bump:

Metal now has C++ headers to directly call all features of the Metal API without 3rd party integrations or tooling:

https://developer.apple.com/metal/cpp/

Secondly, wrt to the future of Metal, I'd feel pretty confident its sticking around for some time. Apple has dedicated GPU chipset support, and drivers supporting a long rung of iOS and now M series chipsets with native Metal / SPIRV implementations of the TBDR.

**Comment by Sandvoxel** [comment]
Small bump:
This would be a really nice feature.

**Comment by vade** [comment]
@Sandvoxel FWIW, if you want some fast GPU code for matrix / tensor MLX has Swift bindings now. Might be of use?

**Comment by Sandvoxel** [comment]
> @Sandvoxel FWIW, if you want some fast GPU code for matrix / tensor MLX has Swift bindings now. Might be of use? 

Thanks for the info but unfortunately I'm using rust.

**Comment by vade** [comment]
https://crates.io/crates/mlx-sys :)
